src/Polly.Core/Timeout/TimeoutResilienceStrategy.cs

```diff
@@ -0,0 +1,108 @@
+using System;
+using System.Threading.Tasks;
+using Polly.Telemetry;
+
+namespace Polly.Timeout;
+
+internal class TimeoutResilienceStrategy : ResilienceStrategy
+{
+    private readonly TimeProvider _timeProvider;
+    private readonly ResilienceTelemetry _telemetry;
+    private readonly TimeoutStrategy _strategy;
+    private readonly Func<TimeoutGeneratorArguments, ValueTask<TimeSpan>> _timeoutGenerator;
+    private readonly Func<OnTimeoutArguments, ValueTask> _onTimeout;
+
+    public TimeoutResilienceStrategy(TimeoutStrategyOptions options, TimeProvider timeProvider, ResilienceTelemetry telemetry)
+    {
+        _timeoutGenerator = options.TimeoutGenerator.CreateHandler();
+        _onTimeout = options.OnTimeout.CreateHandler();
+        _timeProvider = timeProvider;
+        _telemetry = telemetry;
+        _strategy = options.TimeoutStrategy;
+    }
+
+    protected internal override async ValueTask<TResult> ExecuteCoreAsync<TResult, TState>(Func<ResilienceContext, TState, ValueTask<TResult>> callback, ResilienceContext context, TState state)
+    {
+        var timeout = await _timeoutGenerator(new TimeoutGeneratorArguments(context)).ConfigureAwait(context.ContinueOnCapturedContext);
+        if (timeout == System.Threading.Timeout.InfiniteTimeSpan || timeout <= TimeSpan.Zero)
+        {
+            // do nothing
+            return await callback(context, state).ConfigureAwait(context.ContinueOnCapturedContext);
+        }
+
+        var previousToken = context.CancellationToken;
+        using var cancellationSource = new CancellationTokenSource();
+        _timeProvider.CancelAfter(cancellationSource, timeout);
+        context.CancellationToken = cancellationSource.Token;
+
+        CancellationTokenRegistration? registration = null;
+
+        if (previousToken.CanBeCanceled)
+        {
+            registration = previousToken.Register(static state => ((CancellationTokenSource)state!).Cancel(), cancellationSource, useSynchronizationContext: false);
```

Comment:
Is it safe to assume `state` is always a `CancellationToken` here?

Reply:
yup, will be always `CancellationTokenSource`